sections/testing/unit-testing.tex

```diff
@@ -0,0 +1,97 @@
+\section{Unit testing}
+Unit testing is our primary way of testing our system, and belongs in the dynamic white-box testing category.
+The purpose of unit testing is to see if an isolated subset of the system works as expected, given some pre-defined input \cite{TestDrivenDevelopment}.
+For this project, unit tests are utilized for both the React frontend and the API, to ensure that the system works as intended.
+
+\subsection{Choice of framework}
+One of the first decisions when creating unit tests, is the choice of test runner.
+Since the React library is shipped with Jest \cite{ReactRunningTests}, and since both parts of the system are JavaScript based, it seemed like the optimal solution to use the same test runner.
+However, after using Jest for testing the API for a while, several issues surfaced when trying to implement mock functions related to our ORM and some TypeScript functionality in general.
+To simplify the creation of unit testing on the API, the test driver for this was changed to Mocha.
+
+One of the main differences between Jest and Mocha is that Jest is a complete suit that supports running the tests, asserting results, stubbing, test coverage and much more.
+Mocha, on the other hand, is just a test runner and requires additional libraries for asserting, stubbing and generating code coverage results \cite{Mocha}.
+We chose to use Chai for assertions, Sinon for stubbing and Nyc for generating code coverage.
+
+\subsection{What to test}
+The answer to what should be tested varies a lot between the frontend and the API.
+One of the axioms of testing states that it is impossible to test a program completely, and should be seen as a risk-based exercise where the tester asserts which parts of the system should be tested, and how thoroughly it should be tested \cite{SoftwareTesting}.
+For the frontend of the system, it would be possible to test whether each component is rendered to the user, if all the information is correct, what happens if you use 0 as input to the function, 1 as input, 3 as input.. And so on.
```

Comment:
```suggestion
For the frontend of the system, it would be possible to test whether each component is rendered to the user, if all the information is correct, what happens if you use 0 as input to the function, 1 as input, 3 as input etc.
```